src/OpenTelemetry.Extensions.Hosting/README.md

```diff
@@ -9,6 +9,99 @@
+## Usage
+
+### Tracing
+
+#### Simple Configuration
+
+The following example registers tracing using the `ZipkinExporter` and binds
+options to the "Zipkin" configuration section:
+
+```csharp
+services.AddOpenTelemetryTracing((builder) => builder
+    .AddAspNetCoreInstrumentation()
+    .AddHttpClientInstrumentation()
+    .AddZipkinExporter());
+
+services.Configure<ZipkinExporterOptions>(this.Configuration.GetSection("Zipkin"));
+```
+
+#### Using Dependency Injection
+
+The following example registers a processor of the type "MyProcessor" which has
+been registered as a singleton with the `IServiceCollection`:
+
+```csharp
+services.AddSingleton<MyProcessor>();
+
+services.AddOpenTelemetryTracing((builder) => builder
+    .AddAspNetCoreInstrumentation()
+    .AddHttpClientInstrumentation()
+    .AddProcessor<MyProcessor>());
+```
+
+Similar methods exist for registering instrumentation (`AddInstrumentation<T>`)
+and setting a sampler (`SetSampler<T>`).
+
+You can also access the application `IServiceProvider` directly and accomplish
+the same registration using the `Configure` extension like this:
+
+```csharp
+services.AddSingleton<MyProcessor>();
+
+services.AddOpenTelemetryTracing(hostingBuilder => hostingBuilder
+    .Configure((sp, builder) => builder
+        .AddAspNetCoreInstrumentation()
+        .AddHttpClientInstrumentation()
+        .AddProcessor(sp.GetRequiredService<MyProcessor>())));
+```
+
+**Note:** `Configure` is called _after_ the `IServiceProvider` has been built
+from the application `IServiceCollection` so any services registered in the
+`Configure` callback will be ignored.
+
+#### Building Extension Methods
+
+Library authors may want to configure the OpenTelemetry `TracerProvider` and
+register application services to provide more complex features. This can be
+accomplished concisely by casting the `TracerProviderBuilder` into an
+`IDeferredTracerProviderBuilder` instance in an extension method like this:
+
+```csharp
+public static class MyLibraryExtensions
+{
+    public static TracerProviderBuilder AddMyFeature(this TracerProviderBuilder tracerProviderBuilder)
+    {
+        if (!(tracerProviderBuilder is IDeferredTracerProviderBuilder deferredTracerProviderBuilder))
+        {
+            throw new NotSupportedException(
+                "MyFeature requires an IDeferredTracerProviderBuilder instance.");
+        }
+
+        deferredTracerProviderBuilder.Services
+            .AddHostedService<MyHostedService>()
+            .AddSingleton<MyService>()
+            .AddSingleton<MyProcessor>()
+            .AddSingleton<MySampler>();
+
+        return tracerProviderBuilder
+            .AddProcessor<MyProcessor>()
+            .SetSampler<MySampler>();
+    }
+}
+```
+
+Such an extension method can be consumed like this:
+
+```csharp
+services.AddOpenTelemetryTracing((builder) => builder
+    .AddAspNetCoreInstrumentation()
+    .AddHttpClientInstrumentation()
+    .AddZipkinExporter()
+    .AddMyFeature());
```

Comment:
Does this mean ZipkinExporter will run before MyProcessor?

Reply:
> Does this mean ZipkinExporter will run before MyProcessor?

Yes.

I didn't think of this until right now, but this is a kind of interesting situation...

```csharp
.AddMyFeature()
.AddProcessor(new CustomProcessor())
```

In that case CustomProcessor will run _before_ MyFeature's MyProcessor because MyFeature is using `Configure` which defers execution (AddProcessor call) until Build is called.

May or may not be an issue?

Reply:
PS: I switched the order in the README to make it more real-world.

Reply:
Not an issue in my opinion, but some folks might get shocked if they were expecting different behavior.

Basically we need to set the expectation that these are like [TSRs](https://en.wikipedia.org/wiki/Terminate_and_stay_resident_program) and users should understand that ordering does matter.